openlibrary/core/observations.py

```diff
@@ -1,26 +1,245 @@
+        """
+        Returns a dictionary of observation types, values, and IDs.  An observation
+        type/value tuple is the key, and the ID is the value.
+
+        return: Dictionary of observation types, values, and IDs
+        """
+        results = {}
+        for observation in cls.get_observation_types_and_values():
+            results[observation['type'], observation['value']] = observation['value_id']
+
+        return results
+
+    @classmethod
+    def get_patron_observations(cls, username, work_id=None):
+        """
+        Gets all of a patron's observations by default.  Returns only the observations for
+        the given work if work_id is passed.
+
+        return: A list of a patron's observations
+        """
+        oldb = db.get_db()
+        data = {
+            'username': username,
+            'work_id': work_id
+        }
+        query = """
+            SELECT 
+                observation_types.type, 
+                observation_values.value,
+                observation_values.id
+            FROM observations
+            JOIN observation_values
+                ON observations.observation_id = observation_values.id
+            JOIN observation_types 
+                ON observation_values.type = observation_types.id
+            WHERE observations.username = $username"""
+        if work_id:
+            query += " AND work_id=$work_id"
+
+        return list(oldb.query(query, vars=data))
+
+    @classmethod
+    def persist_observations(cls, username, work_id, observations, edition_id=NULL_EDITION_VALUE):
+        """
+        Insert or update a collection of observations.  If no records exist
+        for the given work_id, new observations are inserted.
+
+        """
+
+        def get_observation_ids(observations):
+            """
+            Given a list of observation key-value pairs, returns a list of observation IDs.
+
+            return: List of observation IDs
+            """
+            observation_dict = cls.get_observation_dictionary()
+            observation_ids = []
+
+            for observation in observations:
+                key = list(observation)[0]
+                observation_ids.append(observation_dict[key, observation[key]])
+
+            return observation_ids
+
+
+
+        oldb = db.get_db()
+        records = cls.get_patron_observations(username, work_id)
+
+        if len(records):
+            for record in records:
+                # Delete values that are in existing records but not in submitted observations
+                if { record['type']: record['value'] } not in observations:
+                    observation_id = record['id']
+                    cls.remove_observations(username, work_id, edition_id=edition_id, observation_id=observation_id)
+                else:
+                    # If same value exists in both existing records and observations, remove from observations
+                    observations.remove({ record['type']: record['value'] })
```

Comment:
Should we update the `observations.updated` timestamp here?  If not, that field can be removed from the table altogether.

Reply:
I agree with its removal